### DISCUSSION TITLE
ORM model self-referencing relation (parent / children)

### DISCUSSION THREAD

**Discussion by jacobclarke92** [marcj/deepkit#114] [Q&A] [answered]
I've been trying to work out how to make a self-referencial model, as the title suggest for the purposes of a parent / children structure.

i.e.
```typescript
@entity.name('page').collectionName('pages')
export class Page {
   @t.primary.autoIncrement public id: number = 0
   @t.number private parentId?: number
   // @t.reference({ foreignKey: 'parentId' ) public parent: Page
   // @t.array(() => Page).backReference() public children?: Page[]
}
```

I realise the parentId field is doing nothing in the above example and the references are completely invalid but I'm used to being able to specify an association with a foreign key and couldn't find anything in the docs or type definition comments on how to achieve something like this. 
I saw the mappedBy option in backReference but couldn't find any example of how to use it correctly.
Is it possible without making use of a join table?

Cheers!

**Comment by marcj** [comment, marked as answer]
You would specify the entity like that:

```typescript
@entity.name('page').collectionName('pages')
export class Page {
   @t.primary.autoIncrement public id: number = 0
   @t.reference() public parent?: Page;

   @t.array(() => Page).backReference() public children?: Page[];
}
```

This would create 

```sql
CREATE TABLE `pages` (
        `id` double NOT NULL AUTO_INCREMENT,
        `parent` double NULL,
        PRIMARY KEY (`id`),
        CONSTRAINT `pages_fkb8c9730096d221d4`
            FOREIGN KEY (`parent`)
            REFERENCES `pages` (`id`) ON UPDATE CASCADE ON DELETE CASCADE
    )
    CREATE  INDEX `pages_fkb8c9730096d221d4` ON `pages` (`parent`)
```

**Reply by jacobclarke92** [reply]
Thanks! Seeing the raw SQL in that's generated from the schema is quite helpful. I'll give it a go, cheers!
